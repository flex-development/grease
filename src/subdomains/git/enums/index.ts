/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
