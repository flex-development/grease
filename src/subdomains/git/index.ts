/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './options'
export * from './providers'
export * from './queries'
export type * from './types'
