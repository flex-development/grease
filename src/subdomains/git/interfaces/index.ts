/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
