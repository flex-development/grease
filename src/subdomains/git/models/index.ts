/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
