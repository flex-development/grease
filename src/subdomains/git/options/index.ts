/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitOptions } from './git.options'
