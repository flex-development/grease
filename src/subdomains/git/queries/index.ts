/**
 * @file Entry Point - Queries
 * @module grease/git/queries
 */

export { default as CommitQueryHandler } from './commit.handler'
export { default as CommitQuery } from './commit.query'
export { default as TagQueryHandler } from './tag.handler'
export { default as TagQuery } from './tag.query'
