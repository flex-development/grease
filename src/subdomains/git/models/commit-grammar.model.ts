/**
 * @file Models - CommitGrammar
 * @module grease/git/models/CommitGrammar
 */

import { CommitGrammarOptions } from '#src/git/options'
import {
  join,
  regexp,
  select,
  template,
  unique,
  type Partial
} from '@flex-development/tutils'

/**
 * Commit parser grammar.
 *
 * @class
 */
class CommitGrammar {
  /**
   * Grammar options.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitGrammarOptions} options
   */
  protected readonly options: CommitGrammarOptions

  /**
   * Create a new commit grammar instance.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @param {Partial<CommitGrammarOptions>?} [options] - Grammar options
   */
  constructor(options?: Partial<CommitGrammarOptions>) {
    this.options = new CommitGrammarOptions(options)
  }

  /**
   * Regular expression matching fields in a raw commit.
   *
   * @see https://regex101.com/r/QnFMsV
   *
   * @public
   *
   * @return {RegExp} Raw commit field regex
   */
  public get field(): RegExp {
    return /^-(?<field>[^\s$]+?)-\n(?<value>.*?(?=\n*^-\S+?-)|(?:.*$))/gms
  }

  /**
   * Regular expression matching a commit header.
   *
   * @see https://regex101.com/r/WAJeLp
   * @see https://conventionalcommits.org
   *
   * @public
   *
   * @return {RegExp} Commit header regex
   */
  public get header(): RegExp {
    return /^(?<type>[a-z]+)(?:\((?<scope>[a-z-]+)\))?(?<breaking>!)?: +(?<subject>(?:.+ \([!#](?<pr>\d+)\))|.+)/i
  }

  /**
   * Regular expression matching mentions (e.g. `@unicornware`) in a raw commit.
   *
   * @see https://regex101.com/r/upbRpj
   *
   * @public
   *
   * @return {RegExp} Mention regex
   */
  public get mention(): RegExp {
    return /\B(?<mention>@(?<user>[\w-]{1,38}(?=[^\w-]\B|\s)))/g
  }

  /**
   * Get a regular expression matching references in a raw commit.
   *
   * @see https://regex101.com/r/Thsp1M
   *
   * @public
   *
   * @return {RegExp} Reference regex
   */
  public get reference(): RegExp {
    /**
     * Pattern matching issue references in commit bodies.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '(?:(?:(?<action>{actions}) +)|(?<repository>(?<owner>[\\da-z](?:-(?=[\\da-z])|[\\da-z]){0,38}(?<=[\\da-z]))\\/(?<repo>\\S+)))?(?<ref>(?<prefix>{prefixes})(?<number>\\d+))'

    return new RegExp(
      template(pattern, {
        actions: join(this.options.actions, '|'),
        prefixes: join(select(unique([
          ...this.options.issues,
          ...this.options.pr
        ]), null, regexp), '|')
      }),
      'gi'
    )
  }

  /**
   * Regular expression matching a semantic release tag.
   *
   * @see https://regex101.com/r/WyQyHZ
   * @see https://semver.org
   *
   * @public
   *
   * @return {RegExp} Semantic release tag regex
   */
  public get tag(): RegExp {
    /**
     * Pattern matching semantic release tags.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '^(?<prefix>{prefix})(?<version>(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<prerelease>(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*))*))?(?:\\+(?<metadata>[\\dA-Za-z-]+(?:\\.[\\dA-Za-z-]+)*))?)$'

    return new RegExp(template(pattern, { prefix: this.options.tagprefix }))
  }

  /**
   * Regular expression matching trailers in a raw commit.
   *
   * @see https://git-scm.com/docs/git-interpret-trailers
   * @see https://regex101.com/r/I56Kgg
   *
   * @return {RegExp} Git trailer regex
   */
  public get trailer(): RegExp {
    return /(?<=^|\n)(?<token>[\w -]+?(?=:)): (?<value>[\S\s]+?(?:(?=\n[\w -]+?(?=:))|(?=\n*$)))/g
  }
}

export default CommitGrammar
