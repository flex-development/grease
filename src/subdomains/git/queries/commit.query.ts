/**
 * @file Queries - CommitQuery
 * @module grease/git/queries/CommitQuery
 */

import { IsConstructor } from '#src/decorators'
import { Commit } from '#src/git/models'
import { GitOptions } from '#src/git/options'
import type { CommitConstructor } from '#src/git/types'
import {
  get,
  isArray,
  isString,
  select,
  type Constructor,
  type Partial
} from '@flex-development/tutils'
import { IsArray, IsString, MinLength } from 'class-validator'

/**
 * Git commits query.
 *
 * @see {@linkcode Commit}
 * @see {@linkcode GitOptions}
 *
 * @template T - Parsed commit type
 *
 * @class
 * @extends {GitOptions}
 */
class CommitQuery<T extends Commit = Commit> extends GitOptions {
  /**
   * Parsed commit class.
   *
   * @default Commit
   *
   * @public
   * @instance
   * @member {CommitConstructor<T>} Commit
   */
  @IsConstructor()
  public Commit: CommitConstructor<T>

  /**
   * Reference action keywords.
   *
   * @default
   *  based on repository provider
   *
   * @public
   * @instance
   * @member {string[]} actions
   */
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  public actions: string[]

  /**
   * Revision range start.
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {string} from
   */
  @IsString()
  public from: string

  /**
   * Issue reference prefixes.
   *
   * @default
   *  based on repository provider
   *
   * @public
   * @instance
   * @member {string[]} issues
   */
  @IsArray()
  @IsString({ each: true })
  public issues: string[]

  /**
   * Pull request reference prefixes.
   *
   * @default
   *  based on repository provider
   *
   * @public
   * @instance
   * @member {string[]} pr
   */
  @IsArray()
  @IsString({ each: true })
  public pr: string[]

  /**
   * Revision range end.
   *
   * @default 'HEAD'
   *
   * @public
   * @instance
   * @member {string} to
   */
  @IsString()
  public to: string

  /**
   * Create a new options object.
   *
   * @param {Partial<CommitQuery<T>>?} [opts] - Git commit options
   */
  constructor(opts?: Partial<CommitQuery<T>>) {
    super(opts)

    this.Commit = get(opts, 'Commit', <Constructor<T>>Commit)
    this.actions = <string[]>get(opts, 'actions', [])
    this.from = get(opts, 'from', '')
    this.issues = <string[]>get(opts, 'issues', [])
    this.pr = <string[]>get(opts, 'pr', [])
    this.to = get(opts, 'to', 'HEAD')

    isArray(this.actions) && (this.actions = select(this.actions, isString))
    isArray(this.issues) && (this.issues = select(this.issues, isString))
    isArray(this.pr) && (this.pr = select(this.pr, isString))
  }
}

export default CommitQuery
