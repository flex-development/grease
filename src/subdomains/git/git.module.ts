/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { CqrsModule } from '@nestjs/cqrs'
import { TagOperationHandler } from './operations'
import { GitService } from './providers'
import { CommitQueryHandler, TagQueryHandler } from './queries'

/**
 * Git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  imports: [CqrsModule],
  providers: [
    CommitQueryHandler,
    GitService,
    LoggerService,
    TagOperationHandler,
    TagQueryHandler,
    ValidationService
  ]
})
class GitModule {}

export default GitModule
