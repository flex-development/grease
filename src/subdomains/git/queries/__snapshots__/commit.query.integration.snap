// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:git/queries/CommitQuery > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isConstructor": "Commit must be a constructor",
    },
    "property": "Commit",
    "value": [Function Commit],
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isArray": "actions must be an array",
    },
    "property": "actions",
    "value": Set {},
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "from must be a string",
    },
    "property": "from",
    "value": null,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isArray": "issues must be an array",
    },
    "property": "issues",
    "value": Set {},
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isArray": "pr must be an array",
    },
    "property": "pr",
    "value": Set {},
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "to must be a string",
    },
    "property": "to",
    "value": /\\^HEAD/,
  },
]
`;
